### PR TITLE
`build`: rename main → llama-cli, server → llama-server, llava-cli → llama-llava-cli, etc...

### PR DESCRIPTION
**Pull request by ochafik**
This renames binaries to have a consistent `llama-` prefix.

- `main` → `llama-cli`
- *`any-other-example`* → *`llama-any-other-example`*, e.g:
    - `llama-server`,
    - `llama-baby-llama`,
    - `llama-llava-cli`,
    - `llama-simple`
    - ...
- (`rpc-server` = only exception, name unchanged - see [below](https://github.com/ggerganov/llama.cpp/pull/7809#discussion_r1636043600))

Goal is to make any scripts / docs work across all the ways to install (cmake, brew, rpm, nix...).

For RPM-based distros, the cuda & clblast variants' prefix is `llama-cuda-` / `llama-clblast-`.

The [Homebrew formula](https://github.com/ggerganov/llama.cpp/discussions/7668) will be updated separately (`llama` → `llama-cli`).

Note: can install a local build w/:

```bash
cmake -B build -DLLAMA_CURL=1 && cmake --build build -j && cmake --install build
# Uninstall: ( cd build && xargs rm < install_manifest.txt )
```

See discussion below about naming options considered.

### CODE REVIEW

File: server
Comment (ggerganov) [inline server:1]: Accidentally committed binaries

**Review by Vaibhavs10 — COMMENTED** [review]
Thanks for starting this PR, @ochafik! It's an excellent step to normalise usage across distributors.

Once this is merged, we would need to update the brew formula since we currently rename from `main` to `llama` and everything else to `llama-{basename}`.

It might make sense to move allof the packages to `llama-*` name space. WDYT?

(This will make managing the brew formula easier - so we won't need to have multiple if-else statements)

**Comment by HanClinto** [comment]
> This renames binaries to be consistent w/ the [new Homebrew formula](https://github.com/ggerganov/llama.cpp/discussions/7668) & the Nix build.

This feels wise. There is a lot of documentation and workflow that people will need to update to move away from the `main` name, but overall I think this is the best choice for the project long-term.

**Comment by ggerganov** [comment]
> Running out of ideas on how to verify this PR before merging it, @ggerganov would you have any suggestions?

Nope - we can resolve after `merge` if there are any issues with that

File: .devops/main.Dockerfile
Comment (ngxson) [inline .devops/main.Dockerfile:9]: The name of this Dockerfile also need to be changed

**Comment by ngxson** [comment]
>     * `main` → `llama` (CMake target name: `llama-cli`)

Personally I would prefer to use the name `llama-cli` here instead of just `llama`, since it prevent confusing with the `llama` target (see `target_link_libraries` in CMakeLists). In the long run, good naming can save us from headaches.

File: .devops/main.Dockerfile
Comment (ochafik) [reply .devops/main.Dockerfile:9]: Done, thanks

**Comment by ochafik** [comment]
> > ```
> > * `main` → `llama` (CMake target name: `llama-cli`)
> > ```
> 
> Personally I would prefer to use the name `llama-cli` here instead of just `llama`, since it prevent confusing with the `llama` target (see `target_link_libraries` in CMakeLists). In the long run, good naming can save us from headaches.

I'll leave this one for @ggerganov to pick, tbh.

There's a case to stick w/ `llama` as is already in the Homebrew formula, but `llama-run` would be more explicit. But also, why not just... `llama.cpp`?

`llama.cpp` would work great a supercommand as I mentioned [above](https://github.com/ggerganov/llama.cpp/pull/7809#issuecomment-2155233774): `llama.cpp run -p "Hello"`, `llama.cpp serve -fa`, `llama.cpp quantize ...`.

It would require moving `llama.cpp` - the file not the binary - to `src/`, but I think it would:
- feel rad to type (assuming`llama.cpp.exe` is a valid Windows exec name),
- increase "project brand" awareness,
- avoid squatting the `llama` binary name (not sure how Meta will feel about all this otherwise: IANAL but "Llama 3" is referred to as a "Mark" in [their license](https://llama.meta.com/llama3/license/), for instance). **Possibly saving us from headaches in the long run.**

As for the library vs. CLI target name homonymy we could just rename the lib targets w/ a `-lib` suffix if there's any issue down the line (happy to send https://github.com/ochafik/llama.cpp/commit/0425444ae6fa847687fd50d49347bcd1c1e4fd04 as follow up PR for instance), but I don't think this is problematic now.

**Comment by HanClinto** [comment]
> There's a case to stick w/ `llama` as is already in the Homebrew formula, but `llama-run` would be more explicit. But also, why not just... `llama.cpp`?

I really like all the reasons you laid out -- especially because this project is not "llama" itself -- that's the base model. We're the inference engine.

>     * feel rad to type (assuming`llama.cpp.exe` is a valid Windows exec name),

Or perhaps `llamacpp`?

**Comment by p-e-w** [comment]
If such an invasive change is already being considered, isn't it time to rename the project in its entirety to something other than "llama.cpp"?

Llama/LLaMA is a marketing name used by Meta for a specific family of LLMs released by them. But this project has long supported many LLMs that are completely unrelated to the Llama models, both in lineage and in architecture.

At this point the name "llama.cpp" is

1. misleading, since the software is not limited to Llama models.
2. potentially problematic, since one day Meta might decide that they own the word "llama" in relation to LLMs.

What about simply changing **llama** to **llm** everywhere? `llama.cpp` -> `llm.cpp`, `main` -> `llm-cli`, `server` -> `llm-server` etc.

**Comment by mofosyne** [comment]
Not sure I'm a fan of renaming the project if there is already a mindshare for it, not without a strategic plan to make the project renaming process as clear and smooth as possible (in terms of public signaling).

But I think it's worth having a PR ready to go in case we get into trouble due to the naming by meta. So maybe getting all the relevant image banner update etc...

**Comment by ggerganov** [comment]
I don't think we should change the name of the project. For the binary names, `main` -> `llama-cli`, `server` -> `llama-server`, etc. seems good. The Homebrew package would need to update `llama` -> `llama-cli`. In the future, having a `llama` supercommand would be cool

**Comment by julien-c** [comment]
> In the future, having a `llama` supercommand would be cool

100%, would be awesome in terms of DX in my opinion (like the ability to run interactive command line and the http server from the same entry point)

**Comment by Vaibhavs10** [comment]
Agree with keeping the `llama-*` brand. It has quite a lot of following already and makes a lot more sense to keep.

@ochafik - would you make the update to homebrew? I'm also happy to start a PR parallel-y, too.

**Comment by ochafik** [comment]
> I don't think we should change the name of the project. For the binary names, `main` -> `llama-cli`, `server` -> `llama-server`, etc. seems good. The Homebrew package would need to update `llama` -> `llama-cli`. In the future, having a `llama` supercommand would be cool

@ggerganov Thanks, will update this PR today 👌

> Agree with keeping the `llama-*` brand. It has quite a lot of following already and makes a lot more sense to keep.
> 
> @ochafik - would you make the update to homebrew? I'm also happy to start a PR parallel-y, too.

@Vaibhavs10 Happy to send one, something [like this](https://github.com/Homebrew/homebrew-core/compare/master...ochafik:homebrew-core:llama-cpp-rename) + update git ref

> is this sort of like how I can type `git cola` if git cola is installed? So in this case you are saying we would type something like `llama server` instead of `llama-server` and llama would then respectively call llama-server etc...?

@mofosyne  Yes! Either as a separate binary, or linking the “binaries” as libraries as [here](https://github.com/ggerganov/llama.cpp/compare/master...ochafik:llama.cpp:cli)

**Comment by ochafik** [comment]
Updated the branch, my eyes are refusing to inspect any more diffs 😵, PTAL

**Comment by HanClinto** [comment]
Awesome work @ochafik !! These are a LOT of references to update, and I'm going through and finding dangling threads as I am able.

Question to ponder -- should we also update the `examples/main` directory to `examples/cli` (or something similar)? Currently we refer to it as both the "main executable" but also the "llama-cli executable", and it feels a bit ripe for confusion.

**Comment by HanClinto** [comment]
I don't quite understand what this "apps.nix" file is, but it appears to reference some outdated binary names. Do these need to be updated also? I don't understand why the main binary in here is called `llama` instead of `main` though, so I'm not confident updating this without someone else's 👀 .

https://github.com/ochafik/llama.cpp/blob/bins/.devops/nix/apps.nix

```nix

{
  perSystem =
    { config, lib, ... }:
    {
      apps =
        let
          inherit (config.packages) default;
          binaries = [
            "llama"
            "llama-embedding"
            "llama-server"
            "quantize"
            "train-text-from-scratch"
          ];
          mkApp = name: {
            type = "app";
            program = "${default}/bin/${name}";
          };
        in
        lib.genAttrs binaries mkApp;
    };
}
```

**Comment by HanClinto** [comment]
Opened [a small PR](https://github.com/ochafik/llama.cpp/pull/1) to your branch with a few lingering renames that I caught when searching through.

This is a massive effort and you've done a great job, @ochafik !!

**Comment by ochafik** [comment]
> I don't quite understand what this "apps.nix" file is, but it appears to reference some outdated binary names. Do these need to be updated also? 

Yes they do, done.

> I don't understand why the main binary in here is called llama instead of main though, so I'm not confident updating this without someone else's 👀 .

Both the Nix and Homebrew builds were renaming `main` -> `llama`, but we're now hopefully aligning every distro on `llama-cli`.

> Opened [a small PR](https://github.com/ochafik/llama.cpp/pull/1) to your branch with a few lingering renames that I caught when searching through.

Amazing, thanks!! (good catches 😅)

> This is a massive effort and you've done a great job, @ochafik !!

It's just regexp-replaces that have gone a bit out of control, but thanks!

**Comment by ochafik** [comment]
> > @ochafik - would you make the update to homebrew? I'm also happy to start a PR parallel-y, too.

> @Vaibhavs10 Happy to send one, something [like this](https://github.com/Homebrew/homebrew-core/compare/master...ochafik:homebrew-core:llama-cpp-rename) + update git ref

Actually if you're still keen I'd appreciate you starting that PR, I have no clue how to update the bottle signatures, etc.

**Comment by HanClinto** [comment]
A couple nits I missed in [another small PR here](https://github.com/ochafik/llama.cpp/pull/2).

**Review by HanClinto — APPROVED** [review]
This is beyond my scope to feel I can review comprehensively, but as for my part I've downloaded it, compiled it, run various tests and applications and gone through it by hand as well as I can. Looks good from my perspective, but with something this large probably want at least 1 or 2 others to approve as well.

File: examples/rpc/README.md
Comment (rgerganov) [inline examples/rpc/README.md:3]: The `rpc-server` can be used with any `ggml` program and there is nothing llama specific. So I don't think this change is appropriate. Maybe `ggml-rpc-server`?

Comment (HanClinto) [reply examples/rpc/README.md:3]: The `llama-` prefix is appropriate here if we're eventually going to have a `llama` supercommand -- it's more about project structure and ownership than specific functionality. 

It may more appropriate for the RPC program to be moved to the ggml project, but as long as it lives here, it feels like it's appropriate to live under the `llama-` supercommand.

Comment (ochafik) [reply examples/rpc/README.md:3]: Yeah, saying that `llama-*` binaries are installed by the `llama.cpp` (or `llama-cpp`) package makes things regular, it's a simpler mental model. `llama-gguf` and `llama-ls-sycl-device` are in the same bucket, the latter not even having anything to do w/ ggml.

Comment (ochafik) [reply examples/rpc/README.md:3]: (@rgerganov otherwise using the ggml branding for everything was briefly considered in https://github.com/ggerganov/llama.cpp/pull/7809#issuecomment-2156509720, I think conceptually it would have been the most logical, but branding has its own logic haha)

Comment (rgerganov) [reply examples/rpc/README.md:3]: Ideally the `rpc-server` should be an artifact coming from the `ggml` project but right now it is mixed into `llama.cpp`. The proposed `llama-` naming could be OK for the short term but I can't make the final call here.

Comment (ggerganov) [reply examples/rpc/README.md:3]: I agree that `ggml-rpc-server` better reflects the purpose and functionality of the example. It's place is likely better to be in the `ggml` project, but for greater visibility it is currently located in `llama.cpp` so that we can more easily discuss and experiment with potential applications of this new functionality.

On the other hand, I see that it breaks the pattern with the `llama-` prefix. For now, let's keep `rpc-server` as it is and we'll figure out how to improve later

Comment (ochafik) [reply examples/rpc/README.md]: SG, reverted to rpc-server.

Good to merge?

**Review by ggerganov — COMMENTED** [review]
> [!IMPORTANT]  
> Should add a prominent notification at the top of the main README.md that the examples have been renamed

**Comment by ochafik** [comment]
> Important
> 
> Should add a prominent notification at the top of the main README.md that the examples have been renamed

Indeed, done ([preview](https://github.com/ochafik/llama.cpp/blob/bins/README.md))

**Review by mofosyne — APPROVED** [review]

**Review by ggerganov — APPROVED** [review]
We can merge after the CI is green

Make sure that `ggml-ci` passes locally if you haven't done so yet

**Comment by crashr** [comment]
> Important
> 
> Should add a prominent notification at the top of the main README.md that the examples have been renamed

There is a small nasty pitfall left for users who miss that notification: When someone pulls after that commit without make clean before, the old binaries remain and won't get deleted anymore because the aren't the targets in the Makefile anymore.

**Comment by Nexesenex** [comment]
Also,
https://huggingface.co/spaces/ggml-org/gguf-my-repo
might need to be updated with the new name scheme.
"quantize" (And "imatrix", I suppose) are still called by the script, I guess.

**Comment by mirekphd** [comment]
Yes, the server binary name change from `server` to something more specific (like `llama-server` chosen) was long overdue. 

But it is a breaking change (at least I wasn't prepared with enough asterisks:), so it would be nice if it were documented somewhere other than in this PR,

**Comment by HanClinto** [comment]
> THIS! Y'all just wasted a week of my time. I'm sure I'm not alone.

Just added #8257 to try and address some of this -- would removing legacy binaries as part of `make clean` helped in your particular situation?

**Comment by crashr** [comment]
> Would you mind providing your preferred alternative? Would "you all" be less triggering?

Both versions include me but for sure I have not wasted a week of your time.

**Comment by ngxson** [comment]
Personally I don't think there are any bad things about communications in this PR.

Llama.cpp has high velocity development, much like any other open-source project out there, so breaking changes are expected. We already had dedicated section on README to note about this and other breaking changes.

> because I don't "make clean." I don't know what that is, nor is it mentioned in the readme

> If this doesn't change, you will drive away users.

That's the reason why you shouldn't get mad. Llama.cpp is dev-oriented so will not as friendly as friendly GUI apps out there like jan or lmstudio. We assume that people using llama.cpp at least have basic knowledge about compiling. If you don't know something, then now you have a reason to ask & to learn about that, rather than making complaints.

Finally, I think we should give some credits to devs who dedicated their time to work on big changes like this one. It's not just taking energy to write the code, but also to communicate in a professional & friendly way.

**Comment by crashr** [comment]
@oldmanjk You're not entirely wrong. I recognized the problem because I saw the old binary name in a recent screenshot on Reddit and wondered how that could happen. I generally welcome the change to the binary names, but I also don't see any way of preventing someone from falling foul of this. If, for example, the old binaries had been deleted via makefile, some pipeline somewhere would certainly have broken a productive system. Ok, at least one would have noticed it immediately. Ultimately, one should read the release notes before making any changes like compiling a new binary. Yes, very few of us do. But we should.

**Comment by HanClinto** [comment]
@crashr
> I recognized the problem because I saw the old binary name in a recent screenshot on Reddit and wondered how that could happen. I generally welcome the change to the binary names, but I also don't see any way of preventing someone from falling foul of this.

Yeah, this is a bit concerning to me, and I'm trying to think of what more we can do to mitigate this fallout.

I've put some [suggestions here](https://github.com/ggerganov/llama.cpp/pull/8257#issuecomment-2205107141), but I'm currently leaning towards:

> 2) The second option is to add a minimalistic program that builds to the old binary names, but simply outputs a friendly error message that says that the binary names have changed, and to please update scripts and routines to use the new filenames.

There is more discussion in that PR, but I like the idea of overwriting the old binaries (at least for things like `main`, the server, and maybe quantization commands) so that they don't linger around, and provide people a nice breakpoint so that errors don't continue silently.

@oldmanjk 
> I especially want to thank @HanClinto . 

Thanks for the shoutout, though it takes all kinds to move a project of this size forward, and I'm just a very small part. Nobody who works on this project takes breaking changes like this casually, and -- despite how it may appear from another vantage -- I hope you trust me when I say that nothing in this migration change was done recklessly or without consideration (the pages and pages of discussion at the beginning of the PR should illustrate this). I hope you'll give us all credit for that -- it's messy, difficult, and there are a lot of tradeoffs to balance, and sometimes we adapt on the fly (which is what we're doing here).

This name change should tide us over for a while, and hopefully we won't need another one until we move to the supercommand (and that one should be more backwards-compatible).

**Comment by crashr** [comment]
> What am I wrong about?

The communication thing. What else could have be done but write it in big letters at the top?

**Comment by HanClinto** [comment]
> What else could have be done but write it in big letters at the top?

Well, there are a few things, though some are most clear in hindsight.

One suggestion is implemented in #8283 -- please give it a look-see! :)

**Comment by HanClinto** [comment]
> > You're not entirely wrong.
> 
> What am I wrong about?

The personal deprecations and insinuations are getting tiresome. I'm glad that you appreciate my approach to customer service on bug reports, but aspersions cast at the other volunteers on this project (even if it's only by way of complimenting me to the isolation of others) is not helping me or anyone else. I'm a volunteer here, as is most everyone else, and we all need to strive to be professional and gracious.

A bit of grace and professional courtesy go a long way here -- it's a messy situation, and berating other devs with blame-and-shame is not going to help me do anything other than walk away for a while.

Thank you (and others) for the issue reports, but please recognize that continuing in this vein will wear me (and others) thin.

**Comment by HanClinto** [comment]
> Nice censorship. You've lost my respect, "repo owner"

It's getting personal, off-topic, and starting to go in circles. If you want to continue this argument, there are more appropriate venues.

Thank you.

**Comment by HanClinto** [comment]
> ?That post might have been a mistake. Was I censored? I can't even remember. I edited before you responded

It was a repeat of a baited question aimed at Crashr. It's cluttering up this issue which is (currently) linked as the primary source for information about the name migration. Continuing in circles with this argument makes it only more difficult for other users looking for info about the name change to sift through and find the important information. 

Because of the prominence of this PR (it's the top PR linked from the [main README](https://github.com/ggerganov/llama.cpp/blob/master/README.md), amongst other places), anything we can do to keep this thread focused and un-cluttered is important for other users. If you'd like to continue this argument, please move it to a less visible location.

Thank you.

**Comment by hgftrdw45ud67is8o89** [comment]
personally i don't think this change is needed? Because i thought the old scripts or docs is already stable? But definitely good to make it neat.

But,If its going to be renamed.
We might as well use less confusing name it would be better to just prepend "llamacpp" instead of "llama".
as llama commonly refers to the "llama" arch and not "llamacpp". It have the equal ambiguous level of "main" or "run" or "llm" or "mistral".

Did no one noticed this problem?

**Comment by HanClinto** [comment]
> personally i don't think this change is needed? Because i thought the old scripts or docs is already stable? But definitely good to make it neat.
> 
> But,If its going to be renamed. We might as well use less confusing name it would be better to just prepend "llamacpp" instead of "llama". as llama commonly refers to the "llama" arch and not "llamacpp". It have the equal ambiguous level of "main" or "run" or "llm" or "mistral".
> 
> Did no one noticed this problem?

`llamacpp` was [suggested on June 8th](https://github.com/ggerganov/llama.cpp/pull/7809#issuecomment-2156094066), and [discussed further on June 9th](https://github.com/ggerganov/llama.cpp/pull/7809#issuecomment-2156509720). Final decision was made [on June 10th](https://github.com/ggerganov/llama.cpp/pull/7809#issuecomment-2157457255).

I agree with you that `llamacpp` feels better (at least to me), but I'm still quite content with `llama` and I'm in support. Moving away from the `main` name is a good change.

**Comment by hgftrdw45ud67is8o89** [comment]
Thanks  @HanClinto for giving me reference. My bad for not noticing /reading the full convo.
Its fine to keep it as it is. I just naively thought that no one suggested that idea or noticed the flaw.

**Pull request merged**
